src/guiguts/maintext.py

```diff
@@ -703,6 +703,102 @@ def find_matches(
                 start += f"+{count_var.get()}c"
         return matches
 
+    def transform_selection(self, fn) -> None:
+        """Transform a text selection by applying a function or method.
+
+        Args:
+            fn: Reference to a function or method
+        """
+        sel = self.selected_ranges()
+        if not sel:
+            # nothing is currently selected
+            return
+
+        p1 = self.index("sel.first")
+        p2 = self.index("sel.last")
+        txt = self.selected_text()
+
+        self.delete(p1, p2)
+        # apply transform, then insert at start position
+        self.insert(p1, fn(txt))
+
+        # Note to future devs - this transform assumes the new text is the
+        # same length and has the same line breaks as the old text; if that
+        # isn't true, the re-select may be incorrectly positioned.
+        self.do_select(IndexRange(p1, p2))
+
+    def sentence_case_transformer(self, s) -> str:
+        """Text transformer to convert a string to "Sentence case".
+
+        The transformation is not aware of sentence structure; if the
+        input string consists of multiple sentences, the result will
+        likely not be what was desired. This behavior was ported as-is
+        from Guiguts 1.x, but could potentially be improved, at least
+        for languages structured like English.
+
+        To be more specific: if multiple sentences are selected, the
+        first character of the first sentence will be capitalized, and
+        the subsequent sentences will begin with a lowercase letter.
+
+        Args:
+            s: an input string to be transformed
+
+        Returns:
+            A transformed string
+        """
+        # lowercase string, then look for first word character.
+        # DOTALL allows '.' to match newlines
+        m = re.match(r"(\W*\w)(.*)", s.lower(), flags=re.DOTALL)
+        if m:
+            return m.group(1).upper() + m.group(2)
+        else:
+            return s
+
+    def title_case_transformer(self, s) -> str:
+        """Text transformer to convert a string to "Title Case"
+
+        Args:
+            s: an input string to be transformed
+
+        Returns:
+            A transformed string
+        """
+        # A list of words to *not* capitalize. This list should only be used
+        # for English text.
+        exception_words = (
+            "a",
+            "an",
+            "and",
+            "at",
+            "by",
+            "from",
+            "in",
+            "of",
+            "on",
+            "the",
+            "to",
+        )
+
+        def capitalize_first_letter(match):
+            word = match.group()
+
+            # TODO: At the time this method was implemented, GG2 was not aware
+            # of the document language (`::main_lang()` in GG1 terms). When
+            # such support appears in GG2, we should add a test so that
+            # `exception_words` is only checked when the language is `en`.
```

Comment:
Good point, thanks! I've added [an issue](https://github.com/DistributedProofreaders/guiguts-py/issues/134) to support a document-specific language